utils/names.go

```diff
@@ -40,3 +43,27 @@ func ValidateEnvVarName(name string) error {
 
 	return nil
 }
+
+// ValidateLabel validates that the input name is a valid Kubernetes key.
+// Valid label names have two segments: an optional prefix and name, separated by a slash (/).
+// The name segment is required and must be 63 characters or less, beginning and ending with
+// an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and
+// alphanumerics between. The prefix is optional. If specified, the prefix must be a DNS subdomain:
+// a series of DNS labels separated by dots (.), not longer than 253 characters in total, followed
+// by a slash (/).
+// Valid label values must be 63 characters or less (can be empty),
+// unless empty, must begin and end with an alphanumeric character ([a-z0-9A-Z]),
+// could contain dashes (-), underscores (_), dots (.), and alphanumerics between.
+// Label values may also come from the environment and therefore, could be enclosed with {{}}
+// Treat this as a special case.
+func ValidateLabel(name, value string) error {
+	errs := validation.IsQualifiedName(name)
+	if !strings.HasPrefix(value, "{{") {
```

Comment:
The only issue here is that if the eventual value of the environment variable is an invalid label value, we won't catch it. Maybe I should evaluate the environment variable, and try to validate that.

Reply:
+1

Reply:
>Maybe I should evaluate the environment variable, and try to validate that.

but that can be only done in `deploy` right?

Reply:
> > Maybe I should evaluate the environment variable, and try to validate that.
> 
> but that can be only done in `deploy` right?

It can be evaluated with `func config` as well.

```
> ./func config -p viewer
Error: 'func.yaml' config file is not valid:
  label entry #0 with name 'foo' has invalid value when the environment is evaluated: '-badvalue': a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')
```

Reply:
yeah but then it might be re-set to something else latter before `deploy`

Reply:
> yeah but then it might be re-set to something else latter before `deploy`

That would be OK, right? If my func.yaml looks like this

```
labels:
- name: foo
  value: {{ env:FOO }}
```
And in my local environment I have
```
echo $FOO
-bar # this is an invalid label value
```
When I'm using `func` CLI, I think it's preferable to know before I try to deploy that whatever the value is in my environment is invalid. If I then change the value to something valid, subsequent CLI commands will work. What other did you have in mind?

Reply:
>When I'm using func CLI, I think it's preferable to know before I try to deploy that whatever the value is in my environment is invalid.

OK makes sense, but what do we do in opposite case -- somebody unset the var latter, it makes deployment fail automatically (because server won't accept that)?

Reply:
> OK makes sense, but what do we do in opposite case -- somebody unset the var latter, it makes deployment fail automatically (because server won't accept that)?

The func.yaml file is read and validated with each command. So, if later the env var is set to something invalid, when `func deploy` is run, the first thing that happens is the func.yaml file is read and validated.

Reply:
Yeah, agree with Lance on this. I don't think there anything else that we can do.